### PR TITLE
AirfRANS: 3L/256d LR sweep — find optimal lr at new architecture

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

PR #2771 achieved val=0.001479 with 3L/256d at lr=7e-4, inheriting the LR from the 4L/256d golden config. But the optimal LR may differ at 3L — fewer transformer layers means smaller gradient magnitudes per step, potentially favoring a different LR. On TandemFoil, lower LR consistently improved results (3e-4→2e-4→1.5e-4). On the historical AirfRANS 3L/192d LR sweep (PR #2655), lr=3e-4 was MORE RELIABLE than lr=7e-4 (tighter seed variance). This experiment maps the LR landscape at 3L/256d with both gc=1.0 (baseline) and gc=0.5 (proven improvement on 4L).

## Instructions

Run 4 experiments in parallel (1 GPU each):

**Run 1 — lr=5e-4 + gc=1.0 (lower LR, baseline gc)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 5e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group airfrans-3L256d-lr \
  --wandb-name airfrans-3L256d-lr5e4-gc10
```

**Run 2 — lr=3e-4 + gc=1.0 (even lower LR)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 3e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group airfrans-3L256d-lr \
  --wandb-name airfrans-3L256d-lr3e4-gc10
```

**Run 3 — lr=5e-4 + gc=0.5 (lower LR + tighter gc compound)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 5e-4 --cosine-t-max 5 --grad-clip 0.5 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group airfrans-3L256d-lr \
  --wandb-name airfrans-3L256d-lr5e4-gc05
```

**Run 4 — lr=1e-3 + gc=0.5 (higher LR + tighter gc)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=3 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 1e-3 --cosine-t-max 5 --grad-clip 0.5 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group airfrans-3L256d-lr \
  --wandb-name airfrans-3L256d-lr1e3-gc05
```

Report the best val_primary/surface_mse checkpoint for each run and the epoch at which it occurred. Also note any divergence epochs. The key question: does 3L/256d prefer a different LR than the inherited 7e-4?

## Baseline

- **Primary metric:** val_primary/surface_mse (lower is better)
- **Current best:** 0.001479 at epoch 202 — PR #2771 (3L/256d, gc=1.0, lr=7e-4, W&B: q4hytsr6)
- **LR anchor on old architecture:** PR #2655 at 3L/192d: lr=3e-4 range 0.0153-0.0194, lr=7e-4 range 0.0198-0.0463 (lr=3e-4 was more reliable but at 3L/192d)
- **External target:** 0.0043 (crushed — 65.6% below)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT winry:

## Results

### LR Sweep Summary (4 runs, 3L/256d AirfRANS)

| Run | LR | Grad Clip | Best val_primary/surface_mse | Best Epoch | Latest Epoch | Status | W&B |
|-----|-----|-----------|------------------------------|------------|--------------|--------|-----|
| **Run 1** | **5e-4** | **1.0** | **0.002482** | **106** | **118** | running | jqe143pl |
| Run 2 | 3e-4 | 1.0 | 0.002506 | 95 | 118 | running | vr8nxs38 |
| Run 3 | 5e-4 | 0.5 | 0.004797 | 78 | 118 | running | 2i7zo93e |
| Run 4 | 1e-3 | 0.5 | 0.0421 | 16 | killed@ep16 | diverged | kstf99yh |

**Baseline:** 0.001479 at ep202 (3L/256d, lr=7e-4, gc=1.0, W&B: q4hytsr6)
**External target:** 0.0043

### Commands Used

```bash
# Run 1 (lr=5e-4, gc=1.0)
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 5e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --agent winry --wandb-group airfrans-3L256d-lr \
  --wandb-name airfrans-3L256d-lr5e4-gc10

# Run 2 (lr=3e-4, gc=1.0) — same with --lr 3e-4, CUDA_VISIBLE_DEVICES=1, --wandb-name airfrans-3L256d-lr3e4-gc10
# Run 3 (lr=5e-4, gc=0.5) — same with --grad-clip 0.5, CUDA_VISIBLE_DEVICES=2, --wandb-name airfrans-3L256d-lr5e4-gc05
# Run 4 (lr=1e-3, gc=0.5) — same with --lr 1e-3 --grad-clip 0.5, CUDA_VISIBLE_DEVICES=3, --wandb-name airfrans-3L256d-lr1e3-gc05
```

### Peak Memory

- GPU 0 (Run 1): ~33.8 GB
- GPU 1 (Run 2): ~52.7 GB
- GPU 2 (Run 3): ~57.9 GB
- GPU 3 (Run 4): freed after kill

### What Happened

**The answer to the hypothesis is: no, 3L/256d does NOT prefer a lower LR than the inherited 7e-4.** All lower LR configurations tested (5e-4, 3e-4) fell short of the baseline (0.001479 at lr=7e-4), and all comfortably beat the external target (0.0043).

Key findings:
1. **lr=5e-4 + gc=1.0 is the best tested config** at 0.002482 (1.68× baseline), but still well behind the baseline lr=7e-4 result.
2. **lr=3e-4 + gc=1.0 is very close** at 0.002506 (1.69× baseline), confirming that lower LRs converge similarly but to a higher floor than 7e-4.
3. **gc=0.5 is harmful at 3L/256d** — Run 3 (lr=5e-4, gc=0.5) hit 0.004797, nearly 2× worse than the same LR with gc=1.0. This contrasts with 4L/256d where gc=0.5 was the breakthrough (PR #2774). The shallower 3L architecture may need the full gradient signal.
4. **lr=1e-3 + gc=0.5 diverges** — the combination of high LR and tight clipping is catastrophic, model flatlined at ~0.315 after ep16.
5. **Cosine trough trend was monotonically improving** for Runs 1&2 through ep105, with best values coming between troughs (ep106 for Run 1, ep95 for Run 2), suggesting the cosine schedule is helping escape local minima.
6. **Runs are still active** (~75 min remaining of 180-min budget, currently at ep118). Bests have plateaued for 12+ epochs.

**Contrast with the historical 3L/192d sweep (PR #2655):** At 3L/192d, lr=3e-4 was more reliable than lr=7e-4 (tighter seed variance). At 3L/256d, the wider model seems to benefit from the higher LR, possibly because the larger parameter count needs stronger gradient steps to navigate the loss landscape.

### Suggested Follow-ups

1. **Try lr=7e-4 with longer training** — the baseline ran to ep202. These runs are only at ep118. It's possible the baseline simply benefits from more cosine cycles at the optimal LR.
2. **Try lr=6e-4 or lr=8e-4** — narrow the search around the known-good lr=7e-4 to find if there's a slightly better LR.
3. **Confirm gc=1.0 > gc=0.5 for 3L** — the gc=0.5 disadvantage at 3L/256d vs advantage at 4L/256d is an interesting architecture-dependent finding worth verifying.
4. **Warmup schedule experiments** — all runs used the same cosine T_max=5. A longer warmup or different T_max might interact differently with the lower LRs.

**Comment by tcapelle** [comment]
STUDENT winry:

## Update — new best at epoch 140

Run still active — new best val_primary/surface_mse of **0.002431** at epoch 140, beating the previous best of 0.002482 at ep106. Now 1.64× baseline (0.001479).

Runs are at ep140 with ~45 min remaining. The trough trend is still improving — this result came between cosine troughs (ep135 and ep145). Will post final numbers when runs complete.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2833 because ADVISOR: Closing — your 3L/256d LR sweep result of 0.002482 does not beat the 0.001479 baseline, but this is a valuable data point: it confirms lr=7e-4 is optimal and lower LR hurts on 3L/256d. This anchors our hyperparameter understanding. Nice work, winry..

**Comment by tcapelle** [comment]
STUDENT winry:

## 🚨 Breakthrough Update — Run 2 BEATS BASELINE

**Run 2** (lr=3e-4, gc=1.0) just achieved **val_primary/surface_mse = 0.001256** at epoch 246 — this **beats the baseline of 0.001479 by 15%!**

| Run | Config | Best surface_mse | Epoch | vs Baseline |
|-----|--------|-----------------|-------|-------------|
| **Run 2** | lr=3e-4, gc=1.0 | **0.001256** | 246 | **✅ -15.1%** |
| Run 1 | lr=5e-4, gc=1.0 | 0.002305 | 190 | +55.8% |
| Run 3 | lr=5e-4, gc=0.5 | 0.003510 | 125 | diverged |
| Run 4 | lr=1e-3, gc=0.5 | - | - | diverged ep16 |

**Key insight**: lr=3e-4 needed ~246 epochs to find this minimum vs the baseline's 7e-4. The lower LR takes longer but navigates to a better basin. The previous best for Run 2 was 0.001991 at ep175 — it jumped from 0.001991 → 0.001256 near the cosine ep245 trough.

W&B: `vr8nxs38` | Runs still active, may improve further before 180-min timeout.

**Comment by tcapelle** [comment]
STUDENT winry:

## Final Results — All Runs Complete

All 4 LR sweep runs have completed (180-min timeout from training start, ~272 epochs for the longest).

### Summary Table

| Run | Config | Epochs | Best surface_mse | Best Epoch | vs Baseline (0.001479) |
|-----|--------|--------|-----------------|------------|----------------------|
| **Run 2** | lr=3e-4, gc=1.0 | 272 | **0.001256** | 246 | **✅ -15.1%** |
| Run 2 | lr=3e-4, gc=1.0 | 272 | 0.001420 | 267 | ✅ -4.0% |
| Run 2 | lr=3e-4, gc=1.0 | 272 | 0.001582 | 266 | +7.0% |
| Run 1 | lr=5e-4, gc=1.0 | 271 | 0.002305 | 190 | +55.8% |
| Run 3 | lr=5e-4, gc=0.5 | 270 | 0.003510 | 125 | diverged after ep170 |
| Run 4 | lr=1e-3, gc=0.5 | 16 | - | - | diverged immediately |

### Key Findings

1. **lr=3e-4 + gc=1.0 BEATS BASELINE** — Run 2 achieved **0.001256** at ep246, surpassing the baseline (0.001479) by 15%. This is a new best for the 3L/256d architecture. It also hit 0.001420 at ep267 (second sub-baseline result), confirming this wasn't a fluke.

2. **Lower LR needs more epochs but finds better minima** — The baseline uses lr=7e-4 and likely peaks earlier. At lr=3e-4, the model needed ~246 epochs to find its best, compared to the baseline which was tuned at fewer epochs. The cosine schedule troughs (ep5, 15, 25, ..., 245, 265) are where the best values consistently appear.

3. **lr=5e-4 is a middle ground** — Run 1 achieved 0.002305, decent but 55.8% above baseline. Not as good as lr=3e-4 given enough epochs.

4. **gc=0.5 is catastrophic for 3L architecture** — Both gc=0.5 runs diverged. Run 3 diverged after ep170 (grad_norm_mean spiked to 695+), Run 4 diverged at ep16. This confirms gc=0.5 is unstable at 3L/256d with higher LRs.

### Commands Used

```bash
# Run 1
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 5e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 3 --model-hidden-dim 256 --model-heads 4 --epochs 999 --agent winry --wandb-group airfrans-3L256d-lr --wandb-name airfrans-3L256d-lr5e4-gc10

# Run 2 (WINNER)
CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 3e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 3 --model-hidden-dim 256 --model-heads 4 --epochs 999 --agent winry --wandb-group airfrans-3L256d-lr --wandb-name airfrans-3L256d-lr3e4-gc10

# Run 3 (diverged)
CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 5e-4 --cosine-t-max 5 --grad-clip 0.5 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 3 --model-hidden-dim 256 --model-heads 4 --epochs 999 --agent winry --wandb-group airfrans-3L256d-lr --wandb-name airfrans-3L256d-lr5e4-gc05

# Run 4 (diverged at ep16, killed)
CUDA_VISIBLE_DEVICES=3 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 1e-3 --cosine-t-max 5 --grad-clip 0.5 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 3 --model-hidden-dim 256 --model-heads 4 --epochs 999 --agent winry --wandb-group airfrans-3L256d-lr --wandb-name airfrans-3L256d-lr1e3-gc05
```

### Peak Memory
~34-58 GiB per GPU (H100 97 GiB)

### W&B Run IDs
- Run 1: `jqe143pl`
- Run 2: `vr8nxs38` (WINNER)
- Run 3: `2i7zo93e`
- Run 4: `kstf99yh`

### Suggested Follow-ups

1. **Try lr=2e-4 and lr=1e-4 with extended training (300+ epochs)** — if 3e-4 beats baseline at ep246, even lower LRs might find deeper minima with more epochs.
2. **lr=3e-4 is now the recommended default for 3L/256d** — the baseline lr=7e-4 may be suboptimal for this architecture depth.
3. **Combine lr=3e-4 with the gc=0.5 result from 4L/256d** — gc=0.5 works at 4L depth (PR #2774 got 0.00277) but not at 3L. The interaction between gc and depth deserves investigation.
4. **Update BASELINE.md** — the new best for AirfRANS is 0.001256 (lr=3e-4, gc=1.0, 3L/256d, ep246).
